scripts/check_compliance.py

```diff
@@ -216,26 +231,28 @@ def run(self):
         os.environ["KCONFIG_STRICT"] = "y"
 
         try:
-            kconf = kconfiglib.Kconfig()
+            return kconfiglib.Kconfig()
         except kconfiglib.KconfigError as e:
             self.add_failure(str(e))
-            return
-
-        #
-        # Look for undefined symbols
-        #
+            raise EndTest
 
+    def check_undef_within_kconfig(self, kconf):
```

Comment:
Naive question: would it be possible to just make the actual build fail in this case? I mean within kconfiglib.py in the main zephyr repo.
AFAIK check_compliance.py is for either "soft" rules that need to be overridden from time to time, or for checks that are just too difficult or too slow to implement inside the main build itself. In which category is this one?

Reply:
One thing that's different is that you get a pruned version of the Kconfig tree when running samples and the like (via `source`s that have environment variables in them). This test runs with everything included.

Reply:
Oh, I thought this was for the too-long-menu-test at first, where it's not as essential to see the entire Kconfig tree (but still nice).

For the undefined-within-Kconfig test, it's essential to see the entire Kconfig tree. Otherwise, you'll end up with false positives for symbols that are defined in files that aren't parsed.

You get the same thing in the Linux kernel too. A generic symbol that shows up during an X86 build might depend on `!SOME_ARM_SPECIFIC_THING`, which ends up being undefined. That's fine, because undefined symbols evaluate to `n`.

Reply:
>  ... pruned version of the Kconfig tree when running samples and the like (via sources that have environment variables in them).

Thanks I wasn't aware. Can you point at one specific example?

> You get the same thing in the Linux kernel too

I'm confused. When I run `make menuconfig` in the Linux kernel, I (think I) can search for any obscure sub-sub-sub option, even one that requires a cascade of other dependencies and conditions before it becomes available and shows up in the menus (that's probably the most frequent reason I use kconfig's help: find stuff). I don't see how kconfig could do this without parsing the entire Kconfig tree.

Reply:
Search for `$(ARCH)`, `$(SOC_DIR)`, and `$(BOARD_DIR)` in `Kconfig.zephyr` (that's the syntax for referencing environment variables). Those get set before Kconfig runs, in `cmake/kconfig.cmake`.

For Linux, there's `source "arch/$(SRCARCH)/Kconfig"` in `arch/Kconfig`, which includes e.g. `arch/arm/Kconfig`. Try searching for symbols defined in there while building for x86, and you'll see that they're missing.

Reply:
OK, so you can get the same thing in the Linux kernel too, however if excluding Kconfig files there depends only on SRCARCH (and no SOC_DIR, BOARD_DIR, GPIO42_DIR,...) then it's very unlikely a developer will submit a Kconfig change without testing it with the corresponding ARCH.

Now for fun I purposely broke a couple Kconfig files in the Linux kernel and found that its (original, C and unrelated) kconfig implementation generally doesn't seem to care about undefined symbols and other typos - not even mistakes in the Kconfig files that it does parse. Pretty much anything seems to go, even inverted ranges. I haven't found any "checkconfig" feature in the Linux kernel either so I'm afraid the Linux part of the discussion above was basically entirely irrelevant :-/

Reply:
Leaving Linux aside and rephrasing with a bit more knowledge my initial suggestion at the very top of this thread: why not move this `check_compliance.py#KconfigCheck()` code to `ZEPHYR_BASE/script/kconfig/checkconfig2.py` and ask `cmake` to run it every time anyone builds anything? It seems to take a split second to run and anyone making kconfig changes would most certainly appreciate their mistakes getting caught locally long before CI runs and without the round trips.

Reply:
Adds a bit of overhead that people that reconfigure often might not like (1-2 secs on my Core i7 2600k, though it gets faster once the files are cached). One reason I think people might dismiss Kconfiglib is that they'll assume it'll be too slow since it's doing parsing in Python, so I've gotten kinda obsessed with having it speedy (by Python standards). :)

Maybe it can get annoying during development too, like being forced to constantly run a linter on WIP stuff.

I don't rule it out in the future though. None of my code is ever "done". Stuff can always be improved, and I tend to go back to stuff.

Reply:
And yeah, the C tools are actually kinda crap in many ways, and completely unreadable (which led to the crappiness, because people are afraid of touching the code and breaking something).

Reply:
> And yeah, the C tools are actually kinda crap in many ways, and completely unreadable (which led to the crappiness, because people are afraid of touching the code and breaking something).

Small world: you just reminded me of https://lkml.org/lkml/2018/3/1/1191 "Add Kconfig unit tests" and now the name of one of the reviewers seems familiar :-D

Reply:
Yeah, I've done some cleanups/added some comments to the C tools too. Working on that code always feels a bit masochistic though. :)

Reply:
Any showstoppers? Could approve otherwise.